File: 2017/day09.py
# @formatter:off
from os.path import isfile, join as path_join
from sys import path as sys_path
sys_path.insert(1, path_join(sys_path[0], '..'))
file_name = path_join('input', 'day09.txt')
def to_list(mf: object = int, sep='\n'): return [mf(x) for x in open(file_name).read().split(sep) if x]
def to_gen(mf: object = int, sep='\n'): return (mf(x) for x in open(file_name).read().split(sep) if x)
def p1(*args): ans = part1(*args); print(ans); return ans
def p2(*args): ans = part2(*args); print(ans); return ans

if not isfile(file_name):
	from aoc import get_input_file
	get_input_file(session_path=['..', '.env'])
# @formatter:on

from utils import parsefile

stream = parsefile(file_name, None)


def calc_score(stream, count_garbage=False):
	score = 0
	level = 1

	garbage_count = 0
	in_garbage = False

	i = 0
	while i < len(stream):
		if not in_garbage and stream[i] == '{':
			score += level
			level += 1
		elif not in_garbage and stream[i] == '}':
			level -= 1
		elif not in_garbage and stream[i] == '<':
			in_garbage = True
		elif stream[i] == '>':
			in_garbage = False
		elif in_garbage and stream[i] == '!':
			i += 1
		elif in_garbage:
			garbage_count += 1

		i += 1

	if count_garbage:
		return garbage_count
	else:
		return score


def part1():
	return calc_score(stream)


def part2():
	return calc_score(stream, True)


p1()
p2()

File: cpp/2017/day09.cpp
#include "../aocHelper.h"

class Day09 : public BaseDay
{
public:
	Day09() : BaseDay("09") {}

	result_type solve() override
	{
		long long part1 = 0, part2 = 0;

		string stream;

		// parse input
		while (*input != '\n')
		{
			stream += *input;
			input++;
		}

		auto calc_score = [](string& stream)
		{
			int score = 0;
			int level = 1;

			int garbage_count = 0;
			bool in_garbage = false;

			int i = 0;
			while (i < stream.size())
			{
				if (!in_garbage && stream[i] == '{')
				{
					score += level;
					level++;
				}
				else if (!in_garbage && stream[i] == '}')
				{
					level--;
				}
				else if (!in_garbage && stream[i] == '<')
				{
					in_garbage = true;
				}
				else if (stream[i] == '>')
				{
					in_garbage = false;
				}
				else if (in_garbage && stream[i] == '!')
				{
					i++;
				}
				else if (in_garbage)
				{
					garbage_count++;
				}

				i++;
			}

			return pair<int, int>{ score, garbage_count };
		};

		pair<int, int> res = calc_score(stream);

		// part 1
		part1 = res.first;

		// part 2
		part2 = res.second;

		return { part1, part2 };
	}
};
